### ISSUE TITLE
tf.print() with XLA compilation

### ISSUE THREAD

**Issue by MIHIRKHAMBETE** [tensorflow/tensorflow#52944]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

Hello,

Does anyone know whether tf.print (or an equivalent workaround) can be made to work with XLA compilation? I get an error saying that XLA does not recognize the "printV2" operation if I have a tf.print statement inside a function decorated with @tf.function(jit_compile=True).

If this functionality does not exist, would like to request that it be added as a new feature!

Thanks!

**System information**
- TensorFlow version (you are using): 2.6.0
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
tf.print() does not seem to work when used within functions that are compiled with XLA ( decorated with @tf.function(jit_compile=True)

**Will this change the current api? How?**
No

**Who will benefit with this feature?**
Users of XLA-compiled tensorflow code. 

**Any Other info.**

**Comment by sachinprasadhs** [comment]
Could you please provide the code for which you are facing error. Thanks!

**Comment by bhack** [comment]
Why not

```
import tensorflow as tf

@tf.function(jit_compile=True)
def print_ten():
  for i in range(10):
    print('printing ' +str(i))
  return

print_ten()
```

**Comment by sachinprasadhs** [comment]
Normal print functions works without any limitations, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/8140cf768b36a8daa07505687fceb4a1/52944.ipynb). Thanks!

**Comment by MIHIRKHAMBETE** [comment]
The problem is that with normal print, the print statements are only executed on the first call to the function if the function is called multiple times with the same arguments (since tf will use an already made graph, which does not include the calls to print) - (please see https://www.tensorflow.org/guide/function under section "what is tracing?" for details)

tf.print() on the other hand will always print in graph mode (without XLA), even on non-first calls. However, tf.print() is not currently supported when XLA compilation is used.

Hope this clarifies the nature of the problem -please do let me know if you need more info/code examples of the issue.

**Comment by MIHIRKHAMBETE** [comment]
For example,

`import tensorflow as tf

@tf.function(jit_compile=True)
def print_ten():
  for i in range(10):
    print('printing ' +str(i))
  return

print_ten()
print_ten()
`
Will only print 10 times (corresponding to the first call). If the regular print is replaced with tf.print(), the code will NOT run since XLA does not support tf.print at this time.

**Comment by bhack** [comment]
Yes if you want to print on every function call it is not supported. You can print the returned value outside of the compiled function.
I don't know if it is still updated but here you can find the list of the supported ops:
https://github.com/tensorflow/tensorflow/issues/14798#issuecomment-350814888

**Comment by MIHIRKHAMBETE** [comment]
Yup - that's why I made a feature request for adding tf.print() to the supported ops for XLA compilation.

**Comment by wilecoyote2015** [comment]
Is there a technical reason why `tf.print` is not supported using XLA?
I am really missing this, especially for custom training loops.

**Comment by rivershah** [comment]
@sachinprasadhs any updates on this please? `tf.print` does not work with xla 

```Detected unsupported operations when trying to compile graph __inference_run_step_2702[] on XLA_CPU_JIT: PrintV2 (No registered 'PrintV2' OpKernel for XLA_CPU_JIT devices compatible with node```

**Comment by mohantym** [comment]
@MIHIRKHAMBETE !
I am able to replicate this issue with [2.11 version](https://colab.sandbox.google.com/gist/mohantym/8fd86e85b67fc87bc63fbfceee8c14c8/52944.ipynb#scrollTo=Sp-Zpd4OSssL).
Thank you!

**Comment by chenmoneygithub** [comment]
@sachinprasadhs Please help forward this to XLA or ML API team. Debugging graph mode issues often relies on `tf.print`, since XLA graph cannot do `tf.print`, it's extremely hard to debug when I see bugs specific to XLA graph. Thanks!

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by wilecoyote2015** [comment]
There is no mention of `tf.print()` being enabled for XLA in the release notes. As a consequence, the issue is still relevant.

**Comment by Kayyuri** [comment]
Hi @MIHIRKHAMBETE,
Thank you so much for filing this Feature request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!
